### DISCUSSION TITLE
Interface as subtype of Eloquent\Model

### DISCUSSION THREAD

**Discussion by Grldk** [larastan/larastan#1671] [Q&A] [answered]
(This might be more of a general phpstan/generics question, hope you don't mind this question here)

- larastan v2.6.3
- laravel v9.52.9
- phpstan v1.10.20
- phpstan level 6

Currently working to get our codebase error free on phpstan level 6, running into the following:

We have a couple of interfaces used by Eloquent Models, and we use those interfaces as type hints. These type hints generate phpstan errors like `[Interface] in PHPDoc tag @param for parameter $query is not subtype of template type TModelClass of Illuminate\Database\Eloquent\Model`. 

Maybe due to my lack of understanding about generics, but I'm really struggling to find what to best way is to deal with these errors. Is there a way to indicate that this interface is in fact a subtype of `Model`?

One approach I've tried is adding a `* @template T of \Illuminate\Database\Eloquent\Model` in the Interface. This clears up some errors, but I still get errors in functions like this:

```php
    use Illuminate\Database\Eloquent\Collection as EloquentCollection;
    use Illuminate\Database\Eloquent\Model;

    /**
     * @template T of \Illuminate\Database\Eloquent\Model
     */
    interface ModelInterface    

    /**
     * @return null|EloquentCollection<int, ModelInterface<Model>>
     */
    public function getSubjects(): ?EloquentCollection
```
The function returns an EloquentCollection of Models that implement the Interface. But whatever I do, I always end up with a `parameter is not subtype of template type TModel` error. 

Is there a right way to do this?

**Comment by mad-briller** [comment]
can you try to create a reproducer on https://phpstan.org/try please. as this isn't strictly about larastan
this would make it easier for us to interact with it and debug it with you

**Reply by Grldk** [reply]
Thanks for your quick reply! I've tried to make a simple reproduction:

https://phpstan.org/r/cdd8ab68-9f74-4b96-934e-d0bf6b007cc6

The problem I'm dealing with is the error on line 26: `Type  ModelInterface in generic type Collection<int, ModelInterface> in  PHPDoc tag @return is not subtype of template type TModel of Model of  class Collection.`

**Reply by mad-briller** [reply, marked as answer]
you can provide an intersection type which says it will be both a Model and ModelInterface:
https://phpstan.org/r/fe0817de-5bba-4357-80be-f341ebe9184b

**Reply by Grldk** [reply]
That seems to work quite well, thanks very much! I could/should have thought of that...
